aws/rust-runtime/aws-runtime/src/auth/signing_application.rs

```diff
@@ -0,0 +1,277 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::auth;
+use aws_sigv4::http_request::{sign, SignableBody, SignableRequest, SigningParams};
+use aws_smithy_runtime_api::box_error::BoxError;
+use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+
+/// A builder that configures fields for signing and, upon calling `.finalize()`, produces a
+/// signature and adds signing headers to the given request
+///
+/// # Examples
+/// ```rust,no_run
+/// use aws_runtime::auth::signing_application::SigningApplication;
+/// use aws_sigv4::http_request::SigningParams;
+/// # use aws_smithy_runtime_api::box_error::BoxError;
+/// use aws_smithy_runtime_api::client::orchestrator::HttpRequest;
+///
+/// fn signing_params<'a>() -> SigningParams<'a> {
+///     // --snip--
+/// #    todo!()
+/// }
+///
+/// fn request() -> HttpRequest {
+///     // --snip--
+/// #   HttpRequest::empty()
+/// }
+///
+/// # fn example() -> Result<(), BoxError> {
+/// let mut request = request();
+/// let signing_params = signing_params();
+/// SigningApplication::default()
+///     .request(&mut request)
+///     .signing_params(&signing_params)
+///     .finalize()
+/// # }
+/// ```
+#[derive(Debug, Default)]
+pub struct SigningApplication<'a> {
+    request: Option<&'a mut HttpRequest>,
+    signing_params: Option<&'a SigningParams<'a>>,
+    payload_override: Option<SignableBody<'a>>,
+    #[cfg(feature = "event-stream")]
+    deferred_signer_settings: Option<event_stream::DeferredSignerSettings<'a>>,
+}
+
+impl<'a> SigningApplication<'a> {
+    /// Sets the request (required).
+    pub fn request(mut self, request: &'a mut HttpRequest) -> Self {
+        self.set_request(Some(request));
+        self
+    }
+
+    /// Sets the request (required).
+    pub fn set_request(&mut self, request: Option<&'a mut HttpRequest>) -> &mut Self {
+        self.request = request;
+        self
+    }
+
+    /// Sets the [`SigningParams`] for the builder (required).
+    pub fn signing_params(mut self, signing_params: &'a SigningParams<'_>) -> Self {
```

Comment:
We should add an `impl From<Credentials>` for `Identity` where `Credentials` is defined to make these params easier to construct. Would be good to change the SigningParams builder to take `impl Into<Identity>` as an argument as well.

Reply:
Turns out `impl From<Credentials> for Identity` [already exists](https://github.com/smithy-lang/smithy-rs/blob/d42727e3315b0a8f7213c3e5f2e4ccf53b51bbe2/aws/rust-runtime/aws-credential-types/src/credentials_impl.rs#L196-L201). Isn't updating the [SigningParams builder](https://github.com/smithy-lang/smithy-rs/blob/d42727e3315b0a8f7213c3e5f2e4ccf53b51bbe2/aws/rust-runtime/aws-sigv4/src/sign/v4.rs#L148-L149) to take `impl Into<Identity>` a breaking change with respect to `.identity(...)` and `.set_identity(...)`?

Reply:
Ah, I suppose no. `impl Into<Identity>` just makes use of the API more flexible so that the previous clients continue working.

Reply:
Discussed offline that the suggestion is no longer relevant after the introduction of a `SignWith` trait.